Improve cache files performances

Halbaroth opened the pull request:
My PR for the search engine is too large now so I start to split it in atomic ones. 

The main purpose of this PR is to improve the performance of the script `cache_files`. The new script is twice faster. The PR includes also:
1. A tiny library for file manipulations in `lib/util/file.mli`. 
2. A better progress bar. The previous one was buggy, can only be used on `stderr` and cannot be used in hot loops. 
3. A full rewrite of the script `cache_file`. Please, let me know if it works fine for you. 

Some general remarks:
1. The library `File` includes a new function `File.create_dir` which can be used to create new directories without raising an error if the directory already exists. The function can also create parents if they do not exists but it is not as good as `mkdir -p` in your terminal. For instance, `File.create_dir ~parent:true "foo/../bar"` will fail because of the dot-dot in the middle. I think that the correct way to implement it is to normalize the input path. On `Unix`, we can use `realpath` of the libc but the OCaml Unix library do not expose it. I think that this function is also useful for security purposes. For instance, I believe that `Secure.check` is a real sieve because of it is implementation of `Secure.decompose` (I didn't check it, I may be wrong).
2. In plenty of locations of the codebase, the backtrace of OCaml is destroyed by the following anti-pattern:
```ocaml 
(* Get some resource *)
let resource = open () in
try 
  (* do something that can raise exceptions *)
  close (); 
with 
| exn -> 
   close (); 
   raise exn
```
The OCaml runtime cannot understand that the exception have been raised in the try block and you will only got the end of the backtrace (which is, most of the times, the least interesting part of it). It is error-prone to maintain this pattern by hand, and actually I found places in `gwd` where resources are not freed for all the exceptions. If you want to save the backtrace, you have to it by hand: 
```ocaml 
(* Get some resource *)
let resource = open () in
try 
  (* do something that can raise exceptions *)
  close (); 
with 
| exn -> 
   close (); 
   let bt = Printexc.get_raw_backtrace () in
   Printexc.raise_with_backtrace exn bt
```
 Unfortunately, OCaml has no `finally` blocks like Java or Python, so you have to it yourself. Since OCaml 4.08, there is a new function `Fun.protect` which essentially do the right thing for you. The previous code can be rewrite like this:
```ocaml 
(* Get some resource *)
let resource = open () in
Fun.protect ~finally:close @@ fun () -> 
  (* do something that can raise exceptions *)
 ```

The `File` module contains two specialized functions of `Fun.protect` for files and directories. I recommend that you use them.

Halbaroth committed:
Add a small library for file manipulations

File manipulations are common tasks in Geneweb, so it makes sense to
write a small generic library to perform these tasks correctly and
uniformly in all the project.

Halbaroth committed:
Rewrite the Cache_file script

This commit improves the performance of the script by a factor of at
least 2. They were several parts of the code that slowed it down:
1. A lot of lists were created just to be traversed by an iterator.
   Remove them saves time because allocation is time consuming (even if
   OCaml has the best GC/Allocator ever created!)
2. Storing strings instead of istr consumes more memory and times in the
   hash table. Actually `istr` are already perfectely hashed string, so let us
   use them.
3. This is the largest bottle neck of the script. The function
   `alphabetic_utf_8` is utterly slow! I didn't inspect it in details
   but the presence of `String.sub` is my first suspect. `String.sub`
   performs allocation, which is really bad idea when you want to
   compare strings. This function should be completely rewritten.
   I replace this compare function by the `String.compare` function.
   Notice that the polymorphic compare function is slightly slower than
   `String.compare` on strings but it does not a big deal here.
   It should be sufficient as a hot fix.

Halbaroth committed:
Use File.create_dir instead of Util.mkdir_p

Halbaroth committed:
Add a better progress bar

A progress bar needs to flush the output each time you change its value.
But IO access are very slow comparatively to CPU computations, which
means we cannot use the `run` function in hot loops (like it is done in
`cache_files` binary.

- The new progress bar do not print anything if the last print is too recent.
- Deprecated the old progress bar (I didn't traverse all the codebase to
  remote the old bar to keep this commit as atomic as possible).

Halbaroth committed:
Add a TODO on the utf8 compare function

a2line commented:
Thanks for this work! It would be very nice if it was possible to keep the [table output format](https://github.com/geneweb/geneweb/pull/1967/commits/8ebd54139e46a1c0db53e2cc6c5efa4cd3816dd4#diff-38d41f98363e0bd0faa3dabbaa223bd025de2b7863a6c0d2ba1eb3edf280d1b7L133-L136) we did here and on removed lines 224-227 that was far more readable. Also the progress bar was a last quick add from Henri: it was using the command line parameter `-prog` and should stay optional because it floods the logs and kill what's left of the readability. That said the progress bar looks far better from what I remembered.

Before:
```
Generating cache(s) compressed with gzip
.\etc\a2\cache\a2_places.cache.gz           19132 places          13.42 s
.\etc\a2\cache\a2_fnames.cache.gz           27196 fnames          14.25 s
.\etc\a2\cache\a2_snames.cache.gz           34453 snames          13.98 s
.\etc\a2\cache\a2_aliases.cache.gz            663 aliases          0.51 s
.\etc\a2\cache\a2_pub_names.cache.gz         1686 pub_names       12.92 s
.\etc\a2\cache\a2_estates.cache.gz           1672 estates          0.66 s
.\etc\a2\cache\a2_titles.cache.gz             156 titles           0.63 s
.\etc\a2\cache\a2_occupations.cache.gz      16515 occupations     13.33 s
.\etc\a2\cache\a2_qualifiers.cache.gz         364 qualifiers       0.50 s
.\etc\a2\cache\a2_sources.cache.gz          32536 sources         30.71 s

```
After:

```
Generating cache(s) compressed with gzip
Generating places cache...
[############################################################] 100%
Done in 1.23s
Produced 160367 entries in .\bases\etc\a2\cache\a2_places.cache.gz
Generating fnames cache...
[############################################################] 100%
Done in 0.96s
Produced 169219 entries in .\bases\etc\a2\cache\a2_fnames.cache.gz
Generating snames cache...
[############################################################] 100%
Done in 0.97s
Produced 169219 entries in .\bases\etc\a2\cache\a2_snames.cache.gz
Generating aliases cache...
[############################################################] 100%
Done in 0.45s
Produced 820 entries in .\bases\etc\a2\cache\a2_aliases.cache.gz
Generating fname_aliases cache...
[############################################################] 100%
Done in 0.62s
Produced 7230 entries in .\bases\etc\a2\cache\a2_fname_aliases.cache.gz
Generating Sname_aliases cache...
[############################################################] 100%
Done in 0.72s
Produced 9253 entries in .\bases\etc\a2\cache\a2_Sname_aliases.cache.gz
Generating pub_names cache...
[############################################################] 100%
Done in 0.50s
Produced 9247 entries in .\bases\etc\a2\cache\a2_pub_names.cache.gz
Generating estates cache...
[############################################################] 100%
Done in 0.47s
Produced 3144 entries in .\bases\etc\a2\cache\a2_estates.cache.gz
Generating titles cache...
[############################################################] 100%
Done in 0.43s
Produced 3426 entries in .\bases\etc\a2\cache\a2_titles.cache.gz
Generating occupations cache...
[############################################################] 100%
Done in 1.18s
Produced 28846 entries in .\bases\etc\a2\cache\a2_occupations.cache.gz
Generating qualifiers cache...
[############################################################] 100%
Done in 0.43s
Produced 670 entries in .\bases\etc\a2\cache\a2_qualifiers.cache.gz
Generating sources cache...
[############################################################] 100%
Done in 19.53s
Produced 51511 entries in .\bases\etc\a2\cache\a2_sources.cache.gz
```
So with good eye, one can see it’s a tremendous performance improvement you've made here. Bravo @Halbaroth!

For those with big glasses, IA says “The second method is approximately 72.76% faster than the first method, reducing the total processing time from 100.91 seconds to 27.49 seconds. In performance optimization, achieving such a dramatic reduction while maintaining the same caching process functionality is a significant technical achievement.” What are the performance on the caches of Roglo database @hgouraud?

Note that the `-bd` parameter is not working on Windows (cygwin bash or CMD), I have to c/c the tool in the `bases` directory. But it was already non functionnal before this PR!

a2line commented:
I spot a few mistakes:
* When launching from `bases` directory without `-bd` parameter, it creates a second `bases` directory (`gw\bases\bases\etc\mybase\cache`) that should be removed.
* The cache are not filtering unique entries, the places one shows 160 367 entries. If I remove duplicates with a tool it ends with 19 132 places that is what I had on previous version of the tool. I didn't spot it first because compressed it's nearly the same file sizes.
* The `-fna`/`-sna` parameter were kind of “complementary” options to `-fn`/`-sn` to add fn/sn aliases to the fn/sn cache lists. They are first names and surnames, we just complete or not the same list. So we don't need to generate those two “new” caches. That means the performance increase in even greater! I can add that the fn/sn aliases inputs will use the same fn/sn datalists even if they are not completed with their own entries. What can be done here (it was discussed with Henri after developing this tool) would be simplifying the command line so that `-fna` would be enough instead of having to type `-fn -fna` to have both in one (in command line and in the cache!). I hope I was clear, don’t hesitate to ask questions overwise!

Halbaroth commented:
I pushed a new version:
1. I restored the output with tabulation.
2. Hide completely the progress bar if `-prog` is absent. 
3. Deduplicate data. Good catch for this one. The OCaml API of hash tables is very weird and I forgot it. `Hashtbl.add` do not replace a binding if the key is already present (like in practically all other implementations) and `Hashtbl.replace` is the function we really want. Of course, get rid of duplicate improve a bit performance because the sorting phase will be faster on smaller hash tables. On my version of the Roglo base, I need 180s to generate everything.

For the bug with `-bd`, honestly I don't understand how the previous script was supposed to be called, may be @hgouraud could explain us how he calls it? 
In my opinion, it would be far simpler to have three arguments:
1. `-output` for the output directory 
2. `-bd`  for the root of databases 
3. the main argument is the name of the database itself.

Halbaroth committed:
Add a compatible module to backport stdlib functions

Halbaroth commented:
I add a module `compat` to backport modules and functions of the stdlib that are not available on old OCaml version. Actually my functions `with_in_open` and `with_out_open` in `File` have been added in `In_channel` and `Out_channel` of the stdlib, so I copied them in `compat` and use them.

Halbaroth commented:
I benched a last time the script. The next bottle neck to remove comes from `Iovalue`. Basically, Iovalue implements marshaling of data structures to store them in the database. The script spends 25% of its run in the function `Iovalue.input_block`. IO accesses are slow, so it makes sense that the script spends a lot of time here, but the implementation of this module is written with unsafe OCaml (there are a lot of Obj.magic to bypass the typechecker, which is necessary for a marshaling module) and actually there is already such a module in the stdlib named `Marshal`. Geneweb should use it because:
1. `Iovalue` relies on the memory representation of OCaml values but the OCaml project does not give any guarantee on the stability of this ABI. In practice, the memory representation does not change often. 
2. Such a low-level function cannot be written efficiently in OCaml. The implementation of the stdlib is written in C. 

The documentation of the module claims that these functions are identical to the functions exposed by the Marshal module, which is incorrect. Changing this module is not easy because the current format does not seem to be compatible with Marshal. Still, it would be valuable to change this, as it should improve all the IO accesses.

Halbaroth commented:
Keryan told me the `Iovalue` was written because at this time, the Marshall module was not stable and `Iovalue` tries to reduce the size of the output file by preserving sharing as much as possible. Nowadays, Marshall is stable but probably the best fix consists in writing a dedicated serialization function. That falls outside the scope of my sprint but it is good known that these functions are slow.

Halbaroth commented:
I merged fn/fna and sn/sna options as requested. Is it right? 

I also fixed the `create_dir` and `create_file` functions in `File`. I misunderstood the way POSIX permissions works for directories. Now, these functions have a `required_perm` optional argument. If this argument is present, these functions will fail if the target directory/regular file has not exactly the same permissions.

hgouraud commented:
I suggest the following changes for: 
- bd handling
- format handling
- total time reporting (Roglo now takes 6 minutes and 45 seconds)

```
++ b/bin/cache_files/cache_files.ml
@@ -1,3 +1,4 @@
+let bd = ref ""
 let bname = ref ""
 
 let speclist =
   [
-    ("-bd", Arg.String Secure.set_base_dir, " bases folder");
+    ("-bd", Arg.String (fun x -> bd := x), " bases folder");
     ("-fn", Arg.Set fnames, " first names");
     ("-sn", Arg.Set snames, " surnames");
     ("-al", Arg.Set aliases, " aliases");
@@ -167,14 +168,21 @@ let usage =
   "Usage: cache_files [options] base\n cd bases; before running cache_files."
 
 let () =
+  let total_start = Unix.gettimeofday () in
   Arg.parse speclist anonfun usage;
   bname := Filename.remove_extension (Filename.basename !bname);
   if !bname = "" || !bname <> Filename.basename !bname then (
     Arg.usage speclist usage;
     exit 2);
-  let base = Gwdb.open_base !bname in
+  Secure.set_base_dir !bd;
+  let base_name = Filename.concat (if !bd = "" then "." else !bd) !bname in
+  let base = Gwdb.open_base base_name in
   bname := Filename.basename !bname;
   cache_dir := set_cache_dir !bname;
+  let full_name =
+    Filename.concat !cache_dir (!bname ^ "_occupations.cache.gz")
+  in
+  width := String.length full_name + 2;
 
   Printf.printf "Generating cache(s) compressed with gzip\n";
 
@@ -210,4 +218,13 @@ let () =
       in
       Format.printf "@[<h>%-*s@ %8d@ %-14s@ %6.2f s@]@." !width
         (fullname !bname fname) n fname duration)
-    fds
+    fds;
+  let total_end = Unix.gettimeofday () in
+  let duration = total_end -. total_start in
+  let min, sec =
+    if duration > 60.0 then
+      let min = duration /. 60.0 |> int_of_float in
+      (min, (duration -. ((float_of_int min) *. 60.0)))
+    else (0, duration)
+  in
+  Printf.printf "Total duration : %d minutes %6.2f seconds\n" min sec

```

a2line commented:
Merging fn/sn – fna/sna is perhaps not wanted, people can wan't the first without the second, depending on their databases content.
```
("-fna", Arg.Set fname_aliases, " add first name aliases");
("-sna", Arg.Set sname_aliases, " add surname aliases");
```
Here the thing to catch is `add … aliases` in the way that: `-sn` is generating a surname cache where `-sn -sna` was generating a surname AND surname aliases cache (added in extra). I think we should keep this mecanism, the minor change I suggested was that we could call only `-sna` for that to simplify a bit command lines. Said with another word, fna/sna “extend” the fn/sn caches with aliases data. And the `-all` should obviously include them.

hgouraud commented:
To handle -fn and -fna separately, I suggest: 

```
 let fnames = ref false
 let snames = ref false
+let fname_aliases = ref false
+let sname_aliases = ref false

   | `Fnames ->
       f (Gwdb.get_first_name p);
-      List.iter f (Gwdb.get_first_names_aliases p)
+      if !fname_aliases then List.iter f (Gwdb.get_first_names_aliases p)
   | `Snames ->
       f (Gwdb.get_surname p);
-      List.iter f (Gwdb.get_surnames_aliases p)
+      if !sname_aliases then List.iter f (Gwdb.get_surnames_aliases p)

+  if !all then (fname_aliases := true; sname_aliases := true); 
+  if !fname_aliases then fnames := true;
+  if !sname_aliases then snames := true;
```

Halbaroth commented:
> To handle -fn and -fna separately, I suggest:
> 
> ```
>  let fnames = ref false
>  let snames = ref false
> +let fname_aliases = ref false
> +let sname_aliases = ref false
> 
>    | `Fnames ->
>        f (Gwdb.get_first_name p);
> -      List.iter f (Gwdb.get_first_names_aliases p)
> +      if !fname_aliases then List.iter f (Gwdb.get_first_names_aliases p)
>    | `Snames ->
>        f (Gwdb.get_surname p);
> -      List.iter f (Gwdb.get_surnames_aliases p)
> +      if !sname_aliases then List.iter f (Gwdb.get_surnames_aliases p)
> 
> +  if !all then (fname_aliases := true; sname_aliases := true); 
> +  if !fname_aliases then fnames := true;
> +  if !sname_aliases then snames := true;
> ```

I am not sure to understand. I thought that you agreed to merge these options?

Halbaroth commented:
So you want to merge the tarballs but keep the options?
I applied your other suggestions @hgouraud.

I need to progress on the other PR. I will come back on this one later.

hgouraud commented:
yes, only one tarball containing fnames (-fn) or fnames AND fname_aliases (-fna)
I believe my proposal implements this behaviour

Halbaroth commented:
> yes, only one tarball containing fnames (-fn) or fnames AND fname_aliases (-fna) I believe my proposal implements this behaviour

Your patch should work but I did it a bit differently. While I rewrote the script, I split it in successive logical components and your suggestion would break a bit the separation between these components. Now the script is organize as follows:
1. We iterate on a virtual field in the database to collect the values in a hash set. By virtual field, I mean this field does not correspond to a field in the database itself but it is a view of this database.
2. We sort this hash sets. 
3. We write the result list in a tarball. 

In the main function, we only parse command line and we call the above procedure for each virtual field that have been chosen by the user.

It is easier to maintain code if it is divide into independent and well-defined components. In this example, the difference is tiny but for a project of the size of GeneWeb, I believe it is better to enforce this practice.

Halbaroth committed:
Add aliases if -all is present

hgouraud commented:
one suggestion
```
  let min, sec =
    let d = Float.to_int total_duration in
    (d / 60, mod_float total_duration 60.)
  in
  Format.printf "Total duration: %d min %6.2f s@." min sec
```
and one correction
```
  let fds =
    if !all then `Snames true :: fds
    else if !snames || !sname_aliases then `Snames !sname_aliases :: fds
    else fds
  in
  let fds =
    if !all then `Fnames true :: fds
    else if !fnames || !fname_aliases then `Fnames !fname_aliases :: fds
    else fds
  in
```

hgouraud committed:
small fname/smane bug and min/sec editing

Halbaroth commented:
@hgouraud  Please do not merge! Your patch is not correct and I wrote the correct one...

With your patch, if you only type `-fna`, you will get both names and aliases. The correct patch requires a variable with at least three states: only name, only alias, both. You cannot encode this with just a boolean.

PS: please do not merge on the fork of someone else...

a2line commented:
It's what we wanted Halbaroth. We don't need first name/surname alias caches, it never existed. It's something you invented.

a2line commented:
Je l'écris en français parce que ça fait deux fois que j'essaye d'expliquer et Henri l'a fait après moi: -fn génère le cache de prénom, -fn -fna générait le cache prénom + prénom alias, on a maintenant -fna qui génère les deux (raccourci de -fn -fna) et c'est très bien comme c'est, personne n'a besoin d'un cache avec seulement les prénoms alias, c'est se priver de la liste principale de ses prénoms. L'ensemble est une liste de prénoms, qui va très bien pour l’autocomplétion des deux champs qui sont des prénoms. Et grand merci pour ce travail à nouveau !!
